sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveGroupByAll.scala

```diff
@@ -0,0 +1,110 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.analysis
+
+import org.apache.spark.sql.catalyst.expressions.{Attribute, Expression}
+import org.apache.spark.sql.catalyst.expressions.aggregate.AggregateExpression
+import org.apache.spark.sql.catalyst.plans.logical.{Aggregate, LogicalPlan}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.catalyst.trees.TreePattern.{AGGREGATE, UNRESOLVED_ATTRIBUTE}
+
+/**
+ * Resolve "group by all" in the following SQL pattern:
+ *  `select col1, col2, agg_expr(...) from table group by all`.
+ *
+ * The all is expanded to include all non-aggregate columns in the select clause.
+ */
+object ResolveGroupByAll extends Rule[LogicalPlan] {
+
+  val ALL = "ALL"
+
+  /**
+   * Returns true iff this is a GROUP BY ALL aggregate. i.e. an Aggregate expression that has
+   * a single unresolved all grouping expression.
+   */
+  private def matchToken(a: Aggregate): Boolean = {
+    if (a.groupingExpressions.size != 1) {
+      return false
+    }
+    a.groupingExpressions.head match {
+      case a: UnresolvedAttribute => a.name.toUpperCase() == ALL
```

Comment:
nit: `case a: UnresolvedAttribute if a.nameParts.length == 1 => a.nameParts.head.equalsIgnoreCase(ALL)`

Reply:
what's the actual difference between the two? does .name not include the previous parts?

Reply:
`.name` builds a string, which is unnecessary if `nameParts.length > 1`. This doesn't affect correctness though.